state/presence/presence.go

```diff
@@ -380,66 +426,31 @@ type pingInfo struct {
 	Dead  map[string]int64 `bson:",omitempty"`
 }
 
-func (w *Watcher) findAllBeings() (map[int64]beingInfo, error) {
-	session := w.beings.Database.Session.Copy()
-	defer session.Close()
-
-	beingInfos := make(map[int64]beingInfo)
-	beingsC := w.beings.With(session)
-
-	t0 := time.Now()
-	iter := beingsC.Find(bson.M{
-		"_id": bson.M{"$regex": bson.RegEx{"^" + w.modelUUID, ""}},
-	}).Batch(1600).Iter()
-	var being beingInfo
-	for iter.Next(&being) {
-		beingInfos[being.Seq] = being
-	}
-	if err := iter.Close(); err != nil {
-		return nil, errors.Annotate(err, "reading all beings")
-	}
-
-	dt := time.Now().Sub(t0)
-	logger.Debugf("[%s] loaded %d beings in %s", w.modelUUID[:6], len(beingInfos), dt)
-	return beingInfos, nil
-}
-
-// sync updates the watcher knowledge from the database, and
-// queues events to observing channels. It fetches the last two time
-// slots and compares the union of both to the in-memory state.
-func (w *Watcher) sync() error {
-	var allBeings map[int64]beingInfo
-	if len(w.beingKey) == 0 {
-		// The very first time we sync, we grab all ever-known beings,
-		// so we don't have to look them up one-by-one
-		var err error
-		if allBeings, err = w.findAllBeings(); err != nil {
-			return errors.Trace(err)
-		}
-	}
+func (w *Watcher) lookupPings(session *mgo.Session) ([]pingInfo, error) {
 	// TODO(perrito666) 2016-05-02 lp:1558657
 	s := timeSlot(time.Now(), w.delta)
 	slot := docIDInt64(w.modelUUID, s)
 	previousSlot := docIDInt64(w.modelUUID, s-period)
-	session := w.pings.Database.Session.Copy()
-	defer session.Close()
 	pings := w.pings.With(session)
 	var ping []pingInfo
 	q := bson.D{{"$or", []pingInfo{{DocID: slot}, {DocID: previousSlot}}}}
 	err := pings.Find(q).All(&ping)
 	if err != nil && err != mgo.ErrNotFound {
-		return errors.Trace(err)
+		return nil, errors.Trace(err)
 	}
+	return ping, nil
+}
 
+func (w *Watcher) lookForDead(pings []pingInfo) (map[int64]bool, error) {
```

Comment:
thanks. I'm glad it worked out. I generally find a long function hard to rationalize because you have to mentally chunk it, might as well do so in code. :)